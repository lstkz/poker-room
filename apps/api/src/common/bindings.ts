import * as R from 'remeda';
import {
  BaseBinding,
  CreateEventBindingOptions,
  CreateRpcBindingOptions,
} from '../lib';

const bindings = [
  require('../contracts/user/getMe'),
  require('../contracts/user/login'),
  require('../contracts/user/register'),
  require('../contracts/table/createTable'),
  require('../contracts/table/getAllTables'),
  require('../contracts/table/joinTable'),
  require('../contracts/table/leaveTable'),
  require('../contracts/table/getTableById'),
  require('../contracts/table/createTable'),
  require('../contracts/game/getCurrentGame'),
  require('../contracts/game/makeMove'),
  require('../contracts/example/createFoo'),
  require('../contracts/example/getAll'),
];

export function getBindings(type: 'rpc'): CreateRpcBindingOptions[];
export function getBindings(type: 'event'): CreateEventBindingOptions<any>[];
export function getBindings(
  type: 'rpc' | 'event'
): CreateRpcBindingOptions[] | CreateEventBindingOptions<any>[] {
  return R.pipe(
    bindings,
    R.flatMap(obj => Object.values(obj) as BaseBinding<string, any>[]),
    R.filter(x => x.isBinding && x.type === type),
    R.map(x => x.options)
  );
}
