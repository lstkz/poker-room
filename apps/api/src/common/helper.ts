export function renameId<T extends { _id: any }>(
  obj: T
): Omit<T, '_id'> & { id: string } {
  const ret: any = { ...obj };
  ret.id = ret._id.toString();
  delete ret._id;
  return ret;
}
