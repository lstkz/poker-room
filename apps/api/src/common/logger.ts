import { LOG_LEVEL } from '../config';

export const logger = {
  error(...args: any) {
    console.error('ERROR:', ...args);
  },
  debug(...args: any) {
    if (LOG_LEVEL === 'debug') {
      console.log('DEBUG:', ...args);
    }
  },
  info(...args: any) {
    console.log('INFO:', ...args);
  },
};
