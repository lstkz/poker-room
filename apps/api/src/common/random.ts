import crypto from 'mz/crypto';

export async function randomInt() {
  return (await crypto.randomBytes(4)).readUInt32BE(0);
}
