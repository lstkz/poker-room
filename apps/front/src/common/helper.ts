import { AuthData } from 'shared';
import { GlobalActions } from 'src/features/global/interface';
import * as Rx from 'src/rx';
import { setAccessToken } from 'src/services/Storage';
import { RouterActions } from 'typeless-router';

interface HandleAuthOptions {
  Actions: {
    setLoading(isLoading: boolean): any;
    setError(error: string): any;
  };
  auth(): Rx.Observable<AuthData>;
}

export function handleAuth(options: HandleAuthOptions) {
  const { Actions, auth } = options;
  return Rx.concatObs(
    Rx.of(Actions.setLoading(true)),
    Rx.of(Actions.setError('')),
    auth().pipe(
      Rx.mergeMap(authData => {
        setAccessToken(authData.accessToken);
        return [GlobalActions.loggedIn(authData.user), RouterActions.push('/')];
      }),
      Rx.catchError(e => {
        const error = e.response?.error ?? e.message;
        return Rx.of(Actions.setError(error));
      })
    ),
    Rx.of(Actions.setLoading(false))
  );
}
