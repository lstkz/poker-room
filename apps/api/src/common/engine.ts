import { Card } from 'shared';
import { randomInt } from './helper';

export class CardRandomizer {
  private cards: Card[] = [];

  constructor() {
    const allCards = [2, 3, 4, 5, 6, 7, 8, 9, 'T', 'J', 'Q', 'K', 'A'] as const;
    const allColors = ['c', 'd', 'h', 's'] as const;
    allCards.forEach(card => {
      allColors.forEach(color => {
        this.cards.push({ card, color });
      });
    });
  }

  getLeftCards() {
    return this.cards;
  }

  async randomNextCard(): Promise<Card> {
    if (!this.cards.length) {
      throw new Error('No cards left');
    }
    const idx = (await randomInt()) % this.cards.length;
    const ret = this.cards.splice(idx, 1);
    if (!ret.length) {
      return this.randomNextCard();
    }
    return ret[0];
  }
}
