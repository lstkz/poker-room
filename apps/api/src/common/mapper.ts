import * as R from 'remeda';
import { Table } from 'shared';
import { TableModel } from '../collections/Table';
import { UserModel } from '../collections/User';
import { renameId } from './helper';

export function mapTable(
  table: TableModel,
  userMap: Record<string, UserModel>
): Table {
  return {
    ...renameId(table),
    players: table.players
      .map(player => {
        const id = player.userId.toHexString();
        return {
          user: {
            id,
            username: userMap[id].username,
          },
          ...R.pick(player, ['money', 'seat']),
        };
      })
      .sort((a, b) => a.seat - b.seat),
  };
}
