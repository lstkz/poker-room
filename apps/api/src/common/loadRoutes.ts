import * as R from 'remeda';
import { Router } from 'express';
import fs from 'fs';
import Path from 'path';
import { wrapExpress } from './wrapExpress';
import { BadRequestError, UnauthorizedError } from './errors';
import { logger } from './logger';
import { Handler } from '../types';
import { RpcBinding } from '../lib';

const bindings = R.flattenDeep(
  fs.readdirSync(Path.join(__dirname, '../contracts')).map(dir => {
    return fs
      .readdirSync(Path.join(__dirname, '../contracts', dir))
      .map(fileName => {
        const contractModule: Record<
          string,
          RpcBinding
        > = require(`../contracts/${dir}/${fileName}`);
        return Object.values(contractModule)
          .filter(x => x.isBinding && x.type === 'rpc')
          .map(x => x.options);
      });
  })
);

export default function loadRoutes(router: Router) {
  bindings.forEach(options => {
    const actions: Handler[] = [
      (req, res, next) => {
        if (!req.headers.authorization) {
          return next();
        }
        return next(new Error('Authentication not implemented'));
      },
      (req, res, next) => {
        if (options.public) {
          next();
          return;
        }
        if (!req.user) {
          next(new UnauthorizedError('Bearer token required'));
          return;
        }
        next();
      },
    ];
    if (process.env.NODE_ENV !== 'test') {
      logger.info(
        `${options.public ? '[Public]' : '[Auth]'} ${options.signature}`
      );
    }
    actions.push((req, res, next) => {
      if (req.body == null) {
        req.body = {};
      } else if (typeof req.body !== 'object') {
        next(new BadRequestError('Body must be an object'));
        return;
      }
      const params = options.handler.getParams();
      if (options.injectUser) {
        params.shift();
      }
      const diff = R.difference(Object.keys(req.body), params);
      if (diff.length) {
        next(
          new BadRequestError(
            'Following params are not allowed: ' + diff.join(', ')
          )
        );
        return;
      }
      const values = params.map(x => req.body[x]);
      if (options.injectUser) {
        values.unshift(undefined); // todo
      }
      Promise.resolve(options.handler(...values))
        .then(ret => {
          res.json(ret);
        })
        .catch(next);
    });
    router.post('/' + options.signature, wrapExpress(actions) as any);
  });
}
