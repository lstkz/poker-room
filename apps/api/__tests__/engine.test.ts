import * as R from 'remeda';
import { CardRandomizer } from '../src/common/engine';

describe('CardRandomizer', () => {
  it('should return all cards', () => {
    const cr = new CardRandomizer();
    const cards = cr.getLeftCards();
    expect(cards).toHaveLength(52);
    expect(cards).toMatchSnapshot();
  });

  it('should random cards (sync)', async () => {
    const cr = new CardRandomizer();
    const cards = new Set<string>();
    for (let i = 0; i < 52; i++) {
      const card = await cr.randomNextCard();
      cards.add(card.card + card.color);
    }
    expect(cards.size).toEqual(52);
  });

  it('should random cards (async)', async () => {
    const cr = new CardRandomizer();
    const cards = new Set<string>();
    await Promise.all(
      R.range(0, 52).map(async () => {
        const card = await cr.randomNextCard();
        cards.add(card.card + card.color);
      })
    );
    expect(cards.size).toEqual(52);
  });
});
