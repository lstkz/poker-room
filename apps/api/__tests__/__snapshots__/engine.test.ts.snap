// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CardRandomizer should return all cards 1`] = `
Array [
  Object {
    "card": 2,
    "color": "c",
  },
  Object {
    "card": 2,
    "color": "d",
  },
  Object {
    "card": 2,
    "color": "h",
  },
  Object {
    "card": 2,
    "color": "s",
  },
  Object {
    "card": 3,
    "color": "c",
  },
  Object {
    "card": 3,
    "color": "d",
  },
  Object {
    "card": 3,
    "color": "h",
  },
  Object {
    "card": 3,
    "color": "s",
  },
  Object {
    "card": 4,
    "color": "c",
  },
  Object {
    "card": 4,
    "color": "d",
  },
  Object {
    "card": 4,
    "color": "h",
  },
  Object {
    "card": 4,
    "color": "s",
  },
  Object {
    "card": 5,
    "color": "c",
  },
  Object {
    "card": 5,
    "color": "d",
  },
  Object {
    "card": 5,
    "color": "h",
  },
  Object {
    "card": 5,
    "color": "s",
  },
  Object {
    "card": 6,
    "color": "c",
  },
  Object {
    "card": 6,
    "color": "d",
  },
  Object {
    "card": 6,
    "color": "h",
  },
  Object {
    "card": 6,
    "color": "s",
  },
  Object {
    "card": 7,
    "color": "c",
  },
  Object {
    "card": 7,
    "color": "d",
  },
  Object {
    "card": 7,
    "color": "h",
  },
  Object {
    "card": 7,
    "color": "s",
  },
  Object {
    "card": 8,
    "color": "c",
  },
  Object {
    "card": 8,
    "color": "d",
  },
  Object {
    "card": 8,
    "color": "h",
  },
  Object {
    "card": 8,
    "color": "s",
  },
  Object {
    "card": 9,
    "color": "c",
  },
  Object {
    "card": 9,
    "color": "d",
  },
  Object {
    "card": 9,
    "color": "h",
  },
  Object {
    "card": 9,
    "color": "s",
  },
  Object {
    "card": "T",
    "color": "c",
  },
  Object {
    "card": "T",
    "color": "d",
  },
  Object {
    "card": "T",
    "color": "h",
  },
  Object {
    "card": "T",
    "color": "s",
  },
  Object {
    "card": "J",
    "color": "c",
  },
  Object {
    "card": "J",
    "color": "d",
  },
  Object {
    "card": "J",
    "color": "h",
  },
  Object {
    "card": "J",
    "color": "s",
  },
  Object {
    "card": "Q",
    "color": "c",
  },
  Object {
    "card": "Q",
    "color": "d",
  },
  Object {
    "card": "Q",
    "color": "h",
  },
  Object {
    "card": "Q",
    "color": "s",
  },
  Object {
    "card": "K",
    "color": "c",
  },
  Object {
    "card": "K",
    "color": "d",
  },
  Object {
    "card": "K",
    "color": "h",
  },
  Object {
    "card": "K",
    "color": "s",
  },
  Object {
    "card": "A",
    "color": "c",
  },
  Object {
    "card": "A",
    "color": "d",
  },
  Object {
    "card": "A",
    "color": "h",
  },
  Object {
    "card": "A",
    "color": "s",
  },
]
`;
